pkg/filelist/bucket.go

```diff
@@ -122,7 +123,74 @@ func (ls *bucketLister) maybeDecrypt(r io.Reader) ([]byte, error) {
 	return initial, err
 }
 
-func (ls *bucketLister) listFiles(opts ListOpts, cur *blob.ListIterator) ([]File, error) {
+func (ls *bucketLister) listFiles(opts ListOpts, pathPrefix string) ([]File, error) {
+	// Different underlying storage engines will let us scan/glob parts of the bucket differently.
+	var gcsBucket *storage.Client
+	if ls.buck.As(&gcsBucket) {
+		return ls.listFilesFromGCSBucket(opts, pathPrefix)
+	}
+	return ls.listFilesFromCDKBucket(opts, pathPrefix)
+}
+
+func (ls *bucketLister) listFilesFromGCSBucket(opts ListOpts, pathPrefix string) ([]File, error) {
+	var g errgroup.Group
+	datePrefixes := yyyymmdd.Prefixes(opts.StartDate, opts.EndDate)
+
+	discoveredFiles := make(chan []File)
+
+	for _, datePrefix := range datePrefixes {
+		g.Go(func() error {
+			beforeList := func(as func(interface{}) bool) error {
+				var q *storage.Query
+				if as(&q) {
+					q.MatchGlob = fmt.Sprintf("%s/*/%s*/*", pathPrefix, datePrefix)
+				}
+				return nil
+			}
+
+			listOptions := &blob.ListOptions{
+				Prefix:     pathPrefix, //  + "/",
+				BeforeList: beforeList,
+			}
+
+			files, err := ls.listFilesFromCursor(opts, ls.buck.List(listOptions))
+			if len(files) > 0 {
+				go func() {
```

Comment:
I'm curious, why do we run a new goroutine here?

Reply:
I was running into an issue where the channel send wasn't occurring since the channel requires a reading goroutine.